File: typescript/trade/bloxroute.test.ts
import { Keypair, VersionedTransaction } from "@solana/web3.js";
import { bs58 } from "@project-serum/anchor/dist/cjs/utils/bytes";
import { configDotenv } from "dotenv";
configDotenv();

//Create trading transaction using bloxroute

const pk = process.env.PRIVATE_KEY;
const test = async () => {
  try {
    const private_key = pk || "";
    const wallet = Keypair.fromSecretKey(bs58.decode(private_key));
    const mint = "";
    const param = {
      wallet_address: wallet.publicKey.toBase58(), // Your wallet public key
      action: "buy", // "buy" or "sell"
      mint, // contract address of the token you want to trade
      dex: "jupiter", // exchange to trade on. "pumpfun" or "raydium"
      amount: 0.00001, // amount of SOL or tokens
      slippage: 100, // percent slippage allowed
      tip: 0.001, // priority fee
      type: "bloxroute", // "jito" or "bloxroute"
    };

    const url = "https://api.solanaportal.io/api/trading";
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(param),
    });
    if (response.status === 200) {
      // successfully generated transaction
      const data = await response.json();
      const txnBuffer = Buffer.from(data, "base64");
      const txn = VersionedTransaction.deserialize(txnBuffer);

      txn.sign([wallet]);

      const param = {
        encodedTxn: Buffer.from(txn.serialize()).toString("base64"),
      };
      const url = "https://api.solanaportal.io/api/bloxroute";
      const bloxrouteRes = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(param),
      });
      if (bloxrouteRes.status === 200) {
        const signature = (await bloxrouteRes.json()).signature;
        console.log("- txn succeed", "https://solscan.io/tx/" + signature);
      } else {
        console.log("- txn failed, please check the parameters");
      }
    } else {
      console.log(response.statusText); // log error
    }
  } catch (e: any) {
    console.error(e.message);
  }
};

test();

File: python/trade/bloxroute.test.py
#!/usr/bin/env python3
from dotenv import load_dotenv
import os
import base64
import base58
import requests
from solana.keypair import Keypair
from solana.transaction import VersionedTransaction

def main():
    # Load PRIVATE_KEY from .env
    load_dotenv()
    private_key_b58 = os.getenv("PRIVATE_KEY")
    if not private_key_b58:
        raise EnvironmentError("Missing PRIVATE_KEY in environment")

    # Build Keypair
    secret = base58.b58decode(private_key_b58)
    wallet = Keypair.from_secret_key(secret)
    print(f"wallet: {wallet.public_key}")

    # Prepare the trading payload
    param = {
        "wallet_address": str(wallet.public_key),  # Your wallet pubkey
        "action": "buy",                           # "buy" or "sell"
        "mint": "",                                # token mint address
        "dex": "jupiter",                          # "pumpfun", "raydium", etc.
        "amount": 0.00001,                         # amount of SOL or tokens
        "slippage": 100,                           # percent slippage allowed
        "tip": 0.001,                              # priority fee
        "type": "bloxroute",                       # "jito" or "bloxroute"
    }

    # 1) Request an unsigned, partially-signed txn from SolanaPortal
    url_trade = "https://api.solanaportal.io/api/trading"
    resp = requests.post(url_trade, json=param)
    if resp.status_code != 200:
        print("Error fetching transaction:", resp.text)
        return

    # The API returns a base64-encoded VersionedTransaction
    txn_b64 = resp.json()
    txn_bytes = base64.b64decode(txn_b64)

    # 2) Deserialize & sign locally
    txn = VersionedTransaction.deserialize(txn_bytes)
    txn.sign([wallet])

    # 3) Send signed txn to the bloxroute endpoint
    signed_b64 = base64.b64encode(txn.serialize()).decode()
    blox_payload = {"encodedTxn": signed_b64}
    url_blox = "https://api.solanaportal.io/api/bloxroute"
    blox_resp = requests.post(url_blox, json=blox_payload)
    if blox_resp.status_code == 200:
        signature = blox_resp.json().get("signature")
        print("- txn succeed:", f"https://solscan.io/tx/{signature}")
    else:
        print("- txn failed, please check the parameters:", blox_resp.text)

if __name__ == "__main__":
    main()
